packages/flutter/lib/src/painting/inline_linked_text.dart

```diff
@@ -0,0 +1,557 @@
+    if (span is! TextSpan) {
+      return (span, textLinkerMatches, <TapGestureRecognizer>[]);
+    }
+
+    final List<InlineSpan> nextChildren = <InlineSpan>[];
+    final List<TapGestureRecognizer> recognizers = <TapGestureRecognizer>[];
+    List<_TextLinkerMatch> nextTextLinkerMatches = <_TextLinkerMatch>[...textLinkerMatches];
+    int lastLinkEnd = index;
+    if (span.text?.isNotEmpty ?? false) {
+      final int textEnd = index + span.text!.length;
+      for (final _TextLinkerMatch textLinkerMatch in textLinkerMatches) {
+        if (textLinkerMatch.textRange.start >= textEnd) {
+          // Because ranges is ordered, there are no more relevant ranges for this
+          // text.
+          break;
+        }
+        if (textLinkerMatch.textRange.end <= index) {
+          // This range ends before this span and is therefore irrelevant to it.
+          // It should have been removed from ranges.
+          assert(false, 'Invalid ranges.');
+          nextTextLinkerMatches.removeAt(0);
+          continue;
+        }
+        if (textLinkerMatch.textRange.start > index) {
+          // Add the unlinked text before the range.
+          nextChildren.add(TextSpan(
+            text: span.text!.substring(
+              lastLinkEnd - index,
+              textLinkerMatch.textRange.start - index,
+            ),
+          ));
+        }
+        // Add the link itself.
+        final int linkStart = math.max(textLinkerMatch.textRange.start, index);
+        lastLinkEnd = math.min(textLinkerMatch.textRange.end, textEnd);
+        final (InlineSpan nextChild, TapGestureRecognizer recognizer) = textLinkerMatch.linkBuilder(
+          span.text!.substring(linkStart - index, lastLinkEnd - index),
+          textLinkerMatch.linkString,
+        );
+        nextChildren.add(nextChild);
+        recognizers.add(recognizer);
+        if (textLinkerMatch.textRange.end > textEnd) {
+          // If we only partially used this range, keep it in nextRanges. Since
+          // overlapping ranges have been removed, this must be the last relevant
+          // range for this span.
+          break;
+        }
+        nextTextLinkerMatches.removeAt(0);
+      }
+
+      // Add any extra text after any ranges.
+      final String remainingText = span.text!.substring(lastLinkEnd - index);
+      if (remainingText.isNotEmpty) {
+        nextChildren.add(TextSpan(
+          text: remainingText,
+        ));
+      }
+    }
+
+    // Recurse on the children.
+    if (span.children?.isNotEmpty ?? false) {
+      final (
+        Iterable<InlineSpan> childrenSpans,
+        Iterable<_TextLinkerMatch> childrenTextLinkerMatches,
+        Iterable<TapGestureRecognizer> childrenRecognizers,
+      ) = _linkSpansRecurse(
+        span.children!.map((InlineSpan childSpan) => (
+          span: childSpan,
+          length: childSpan.toPlainText().length,
```

Comment:
Since the function is called recursively, for a given span, `toPlainText()` is going to be called the same number of times as it has ancestors?

Reply:
You're right.  I think I can avoid this by creating a Map with all the cached lengths or something like that, I'll work on it.

Reply:
I've cached this in a Map as described, and I moved all of the caching logic into a _TextCache class.